### ISSUE TITLE
Early Stopping on val metrics bugs

### ISSUE THREAD

**Issue by drazellan** [williamFalcon/pytorch-lightning#374]
Hi
I can't use early stopping on monitor='val_loss' or 'val_acc' with gpu_template.py and lightning_module_template.py example code.
I got
 Early stopping conditioned on metric `val_loss` which is not available. Available metrics are: loss <class 'RuntimeWarning'>

I've try to add val_loss in the ouput of validation_end but it doesn't change anythings.

i look at the code and i  print self.callback_metrics i only have {'loss': 2.8375635147094727}

In trainer.py i see that the early stopping callback is call  in the training loop and not in the validation loop 
So if  i got the val_loss key after validation_end in  self.callback_metrics but it's erased when the training loop restart...

Thanks for your work

**Comment by williamFalcon** [comment]
fixed. try again!

**Comment by drazellan** [comment]
thanks!
I try it...

**Comment by drazellan** [comment]
it still does not work...
the same issue again :
Early stopping conditioned on metric `val_loss` which is not available. Available metrics are: loss <class 'RuntimeWarning'>
As i said in my first post, I looked in the dict callback_metrics in trainer.py and there is only the  loss key after the training loop finish, the 'val_loss' key appears in callback_metrics after the validation run but disappears after. I think the issue is in _process_output 
 my code :
"""
Runs a model on a single node across N-gpus.
"""
import os
import numpy as np
import torch

from argparse import ArgumentParser
from pytorch_lightning import Trainer
from lightning_module_template import LightningTemplateModel

SEED = 2334
torch.manual_seed(SEED)
np.random.seed(SEED)

from pytorch_lightning.callbacks import EarlyStopping



def main(hparams):
    """
    Main training routine specific for this project
    :param hparams:
    """
    # ------------------------
    # 1 INIT LIGHTNING MODEL
    # ------------------------
    model = LightningTemplateModel(hparams)

    # ------------------------
    # 2 INIT TRAINER
    # ------------------------
    # trainer = Trainer(max_nb_epochs=1, gpus=[0, 1, 3, 7], distributed_backend='ddp')   

    gpu_param=hparams.gpus
    
    stopping = EarlyStopping(
        monitor='val_loss',
        min_delta=0.00,
        patience=3,
        verbose=False,
        mode='min'
    )
    if hparams.distributed_backend=="ddp":
        gpu_param=[0,1,2,3,4, 5, 6, 7]

    trainer = Trainer(
        gpus=gpu_param,
        distributed_backend=hparams.distributed_backend,
        use_amp=hparams.use_16bit,
        early_stop_callback=stopping,
        min_nb_epochs=-1,
        max_nb_epochs=20,
        val_percent_check=0.1,
        train_percent_check=0.1

    )

    # ------------------------
    # 3 START TRAINING
    # ------------------------
    trainer.fit(model)


if __name__ == '__main__':
    # ------------------------
    # TRAINING ARGUMENTS
    # ------------------------
    # these are project-wide arguments

    root_dir = os.path.dirname(os.path.realpath(__file__))
    parent_parser = ArgumentParser(add_help=False)

    # gpu args
    parent_parser.add_argument(
        '--gpus',
        type=int,
        default=2,
        help='how many gpus'
    )
    parent_parser.add_argument(
        '--distributed_backend',
        type=str,
        default='dp',
        help='supports three options dp, ddp, ddp2'
    )
    parent_parser.add_argument(
        '--use_16bit',
        dest='use_16bit',
        action='store_true',
        help='if true uses 16 bit precision'
    )

    # each LightningModule defines arguments relevant to it
    parser = LightningTemplateModel.add_model_specific_args(parent_parser, root_dir)
    hyperparams = parser.parse_args()

    # ---------------------
    # RUN TRAINING
    # ---------------------
    main(hyperparams)

Thanks

**Comment by williamFalcon** [comment]
did you pull the latest master? (and install the latest master)

**Comment by drazellan** [comment]
I install via the pip command this morning ( in france). 
I'll try with the lastest master tonight or tomorow..

**Comment by drazellan** [comment]
I try with the latest master installed with git.
I don't have a gpu at home but the bug is still here with the cpu_template code with train an val percent set to 0.05
There is only the loss value in the metrics passed to the early_stopping_callback.

**Comment by drazellan** [comment]
If found that there is a __run_training_batch call done after the evaluation is done in run_training_epoch function, 
So the callback_metrics is only fill with the loss value and not the val_loss value.
If i change met_batch_limit = batch_nb > self.nb_training_batches with met_batch_limit = batch_nb >= self.nb_training_batches the problem disappears.
I just noticed it's the same problem as in issue #371
